From 57aa6d00fc4b73f9801ba1d04df78e7399502318 Mon Sep 17 00:00:00 2001
From: Miaoqing Pan <quic_miaoqing@quicinc.com>
Date: Thu, 2 Jun 2022 14:15:27 +0800
Subject: [PATCH 2/4] cfg80211: obtain sae_pwe value from userspace

Let the userspace advertise the value of the SAE mechanism used for PWE
derivation by setting the NL80211_ATTR_SAE_PWE flag.

The hostapd/wpa_supplicant will pass the sae_pwe value using the new
NL80211_ATTR_SAE_PWE attribute in the %NL80211_CMD_CONNECT and
%NL80211_CMD_START_AP to the driver.

The sae_pwe supports three values
0 = hunting-and-pecking loop only
1 = hash-to-element only
2 = both hunting-and-pecking loop and hash-to-element enabled

The driver can use the value of the sae_pwe passed by userspace,
to selectively allow or disallow the usage of few SAE mechanisms
used for PWE derivation.

Change-Id: I46fe03386c3667852f470affca42bd7369bd08a4
Signed-off-by: Rohan <drohan@codeaurora.org>
Signed-off-by: stonez <stonez@codeaurora.org>
Signed-off-by: Miaoqing Pan <quic_miaoqing@quicinc.com>
---
 include/net/cfg80211.h       |  5 +++++
 include/uapi/linux/nl80211.h | 36 +++++++++++++++++++++++++++++++++++-
 net/wireless/nl80211.c       |  5 +++++
 3 files changed, 45 insertions(+), 1 deletion(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index b425cfc3b28c..c1305841e4d7 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -933,6 +933,10 @@ struct survey_info {
  * @sae_pwd: password for SAE authentication (for devices supporting SAE
  *	offload)
  * @sae_pwd_len: length of SAE password (for devices supporting SAE offload)
+ * @sae_pwe: value indicates the SAE mechanism used for PWE derivation
+ *     0 = hunting-and-pecking loop only
+ *     1 = hash-to-element only
+ *     2 = both hunting-and-pecking loop and hash-to-element enabled
  */
 struct cfg80211_crypto_settings {
 	u32 wpa_versions;
@@ -950,6 +954,7 @@ struct cfg80211_crypto_settings {
 	const u8 *psk;
 	const u8 *sae_pwd;
 	u8 sae_pwd_len;
+	u16 sae_pwe;
 };
 
 /**
diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index 04c69d634f2c..7c13e88d75da 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -287,6 +287,17 @@
  * the interface goes down.
  */
 
+/**
+ * DOC: SAE mechanism for PWE derivation
+ *
+ * By setting @NL80211_ATTR_SAE_PWE flag userspace can indicate the SAE
+ * mechanism used for PWE derivation, in WPA3-Personal networks which are
+ * using SAE authentication.
+ *
+ * In %NL80211_CMD_CONNECT and %NL80211_CMD_START_AP the sae_pwe value
+ * for SAE authentication should be specified using the %NL80211_ATTR_SAE_PWE.
+ */
+
 /**
  * enum nl80211_commands - supported nl80211 commands
  *
@@ -350,6 +361,7 @@
  *	%NL80211_ATTR_AKM_SUITES, %NL80211_ATTR_PRIVACY,
  *	%NL80211_ATTR_AUTH_TYPE, %NL80211_ATTR_INACTIVITY_TIMEOUT,
  *	%NL80211_ATTR_ACL_POLICY and %NL80211_ATTR_MAC_ADDRS.
+ *	%NL80211_ATTR_SAE_PWE.
  *	The channel to use can be set on the interface or be given using the
  *	%NL80211_ATTR_WIPHY_FREQ and the attributes determining channel width.
  * @NL80211_CMD_NEW_BEACON: old alias for %NL80211_CMD_START_AP
@@ -600,7 +612,7 @@
  *	%NL80211_ATTR_CONTROL_PORT_ETHERTYPE,
  *	%NL80211_ATTR_CONTROL_PORT_NO_ENCRYPT,
  *	%NL80211_ATTR_CONTROL_PORT_OVER_NL80211, %NL80211_ATTR_MAC_HINT, and
- *	%NL80211_ATTR_WIPHY_FREQ_HINT.
+ *	%NL80211_ATTR_WIPHY_FREQ_HINT, %NL80211_ATTR_SAE_PWE.
  *	If included, %NL80211_ATTR_MAC and %NL80211_ATTR_WIPHY_FREQ are
  *	restrictions on BSS selection, i.e., they effectively prevent roaming
  *	within the ESS. %NL80211_ATTR_MAC_HINT and %NL80211_ATTR_WIPHY_FREQ_HINT
@@ -613,6 +625,9 @@
  *	%NL80211_ATTR_PREV_BSSID can be used to request a reassociation within
  *	the ESS in case the device is already associated and an association with
  *	a different BSS is desired.
+ *	%NL80211_ATTR_SAE_PWE can be used by userspace to indicate the SAE
+ *	mechanism used for PWE derivation in WPA3-Personal networks which are
+ *	using SAE authentication.
  *	Background scan period can optionally be
  *	specified in %NL80211_ATTR_BG_SCAN_PERIOD,
  *	if not specified default background scan configuration
@@ -2465,6 +2480,18 @@ enum nl80211_commands {
  * @NL80211_ATTR_HE_6GHZ_CAPABILITY: HE 6 GHz Band Capability element (from
  *	association request when used with NL80211_CMD_NEW_STATION).
  *
+ * @NL80211_ATTR_RECEIVE_MULTICAST: multicast flag for the
+ *	%NL80211_CMD_REGISTER_FRAME command, see the description there.
+ * @NL80211_ATTR_WIPHY_FREQ_OFFSET: offset of the associated
+ *	%NL80211_ATTR_WIPHY_FREQ in positive KHz. Only valid when supplied with
+ *	an %NL80211_ATTR_WIPHY_FREQ_OFFSET.
+ * @NL80211_ATTR_CENTER_FREQ1_OFFSET: Center frequency offset in KHz for the
+ *	first channel segment specified in %NL80211_ATTR_CENTER_FREQ1.
+ * @NL80211_ATTR_SCAN_FREQ_KHZ: nested attribute with KHz frequencies
+ *
+ * @NL80211_ATTR_SAE_PWE: Indicates the SAE mechanism used for PWE derivation
+ *	in  WPA3-Personal networks which are using SAE authentication.
+ *
  * @NUM_NL80211_ATTR: total number of nl80211_attrs available
  * @NL80211_ATTR_MAX: highest attribute number currently defined
  * @__NL80211_ATTR_AFTER_LAST: internal use
@@ -2942,6 +2969,13 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_HE_6GHZ_CAPABILITY,
 
+	NL80211_ATTR_RECEIVE_MULTICAST,
+	NL80211_ATTR_WIPHY_FREQ_OFFSET,
+	NL80211_ATTR_CENTER_FREQ1_OFFSET,
+	NL80211_ATTR_SCAN_FREQ_KHZ,
+
+	NL80211_ATTR_SAE_PWE,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 752a24b1ec17..ccd3ef1857bb 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -656,6 +656,7 @@ const struct nla_policy nl80211_policy[NUM_NL80211_ATTR] = {
 		.type = NLA_EXACT_LEN,
 		.len = sizeof(struct ieee80211_he_6ghz_capa),
 	},
+	[NL80211_ATTR_SAE_PWE] = { .type = NLA_U16 }
 };
 
 /* policy for the key attributes */
@@ -9397,6 +9398,10 @@ static int nl80211_crypto_settings(struct cfg80211_registered_device *rdev,
 			nla_len(info->attrs[NL80211_ATTR_SAE_PASSWORD]);
 	}
 
+	if (info->attrs[NL80211_ATTR_SAE_PWE]) {
+		settings->sae_pwe =
+			nla_get_u16(info->attrs[NL80211_ATTR_SAE_PWE]);
+	}
 	return 0;
 }
 
-- 
2.25.1

